File: src/FinanceControl.Web/src/lib/currencies.ts
// Currency catalog (MEL-26): mirror of the backend catalog served by GET /api/currencies.
// Every `*_cents` value is an integer amount in MINOR UNITS of the record's own currency (10^decimals).

export type CurrencyKind = "fiat" | "crypto";

export interface CurrencyInfo {
  code: string;
  /** pt-BR name, e.g. "Dólar americano". */
  name: string;
  symbol: string;
  /** Fraction digits of the minor unit (JPY 0, BTC 8). */
  decimals: number;
  kind: CurrencyKind;
}

/** Base currency (fixed in v1.2). */
export const BASE_CURRENCY = "BRL";

export const currencies: readonly CurrencyInfo[] = [
  { code: "BRL", name: "Real brasileiro", symbol: "R$", decimals: 2, kind: "fiat" },
  { code: "USD", name: "Dólar americano", symbol: "US$", decimals: 2, kind: "fiat" },
  { code: "EUR", name: "Euro", symbol: "€", decimals: 2, kind: "fiat" },
  { code: "GBP", name: "Libra esterlina", symbol: "£", decimals: 2, kind: "fiat" },
  { code: "JPY", name: "Iene japonês", symbol: "¥", decimals: 0, kind: "fiat" },
  { code: "CHF", name: "Franco suíço", symbol: "CHF", decimals: 2, kind: "fiat" },
  { code: "CAD", name: "Dólar canadense", symbol: "C$", decimals: 2, kind: "fiat" },
  { code: "AUD", name: "Dólar australiano", symbol: "A$", decimals: 2, kind: "fiat" },
  { code: "ARS", name: "Peso argentino", symbol: "ARS", decimals: 2, kind: "fiat" },
  { code: "BTC", name: "Bitcoin", symbol: "₿", decimals: 8, kind: "crypto" },
  { code: "ETH", name: "Ethereum", symbol: "Ξ", decimals: 8, kind: "crypto" },
  { code: "SOL", name: "Solana", symbol: "SOL", decimals: 8, kind: "crypto" },
  { code: "USDT", name: "Tether (USDT)", symbol: "USDT", decimals: 2, kind: "crypto" },
];

const byCode = new Map(currencies.map(currency => [currency.code, currency]));

/**
 * Catalog entry for `code`; blank → BRL; unknown codes still work (symbol = code, 2 decimals, fiat)
 * so stored values never break the UI.
 */
export function currencyOf(code: string | null | undefined): CurrencyInfo {
  const normalized = String(code ?? "").trim().toUpperCase() || BASE_CURRENCY;
  return byCode.get(normalized) ?? { code: normalized, name: normalized, symbol: normalized, decimals: 2, kind: "fiat" };
}

/** "USD · Dólar americano" */
export const currencyLabel = (code: string | null | undefined): string => {
  const currency = currencyOf(code);
  return currency.name === currency.code ? currency.code : `${currency.code} · ${currency.name}`;
};

/** `[code, label]` options for a plain select: all catalog currencies (fiat first, then crypto). */
export const currencyOptions = (): [string, string][] => currencies.map(currency => [currency.code, currencyLabel(currency.code)]);

/** Options grouped for a select with optgroups: "Moedas" (fiat) and "Cripto". */
export const currencyGroups = (): { label: string; options: [string, string][] }[] => [
  { label: "Moedas", options: currencies.filter(currency => currency.kind === "fiat").map(currency => [currency.code, currencyLabel(currency.code)]) },
  { label: "Cripto", options: currencies.filter(currency => currency.kind === "crypto").map(currency => [currency.code, currencyLabel(currency.code)]) },
];

export const isBaseCurrency = (code: string | null | undefined): boolean => currencyOf(code).code === BASE_CURRENCY;

File: src/FinanceControl.Domain/Rules/Currencies.cs
using System.Globalization;

namespace FinanceControl.Domain.Rules;

/// <summary>Moeda do catálogo: valores guardados em unidades mínimas (10^<see cref="Decimals"/>) da própria moeda.</summary>
public sealed record CurrencyInfo(string Code, string Name, string Symbol, int Decimals, string Kind);

/// <summary>Catálogo único de moedas (v1.2). A moeda base é o real (BRL), com cotação implícita 1.</summary>
public static class Currencies
{
    public const string Base = "BRL";
    public const string Fiat = "fiat";
    public const string Crypto = "crypto";

    public static readonly IReadOnlyList<CurrencyInfo> All =
    [
        new("BRL", "Real brasileiro", "R$", 2, Fiat),
        new("USD", "Dólar americano", "US$", 2, Fiat),
        new("EUR", "Euro", "€", 2, Fiat),
        new("GBP", "Libra esterlina", "£", 2, Fiat),
        new("JPY", "Iene japonês", "¥", 0, Fiat),
        new("CHF", "Franco suíço", "CHF", 2, Fiat),
        new("CAD", "Dólar canadense", "C$", 2, Fiat),
        new("AUD", "Dólar australiano", "A$", 2, Fiat),
        new("ARS", "Peso argentino", "ARS", 2, Fiat),
        new("BTC", "Bitcoin", "₿", 8, Crypto),
        new("ETH", "Ethereum", "Ξ", 8, Crypto),
        new("SOL", "Solana", "SOL", 8, Crypto),
        new("USDT", "Tether (USDT)", "USDT", 2, Crypto)
    ];

    private static readonly Dictionary<string, CurrencyInfo> ByCode = All.ToDictionary(item => item.Code, StringComparer.Ordinal);

    public static bool IsSupported(string? code) => code is not null && ByCode.ContainsKey(code);
    public static CurrencyInfo? Find(string? code) => code is not null && ByCode.TryGetValue(code, out var info) ? info : null;
    public static int DecimalsOf(string code) => Find(code)?.Decimals ?? 2;

    /// <summary>Moedas com cotação (todas menos a base), na ordem do catálogo.</summary>
    public static IEnumerable<CurrencyInfo> Quoted => All.Where(item => item.Code != Base);
}

/// <summary>Conversão para a moeda base: <c>base = round(minor / 10^dec × cotação × 100)</c>, arredondando o meio para longe do zero.</summary>
public static class CurrencyRules
{
    private static readonly CultureInfo Invariant = CultureInfo.InvariantCulture;

    public static long ToBaseCents(long amountMinor, string currency, decimal rateBrl)
    {
        if (currency == Currencies.Base) return amountMinor;
        var scale = Pow10(Currencies.DecimalsOf(currency));
        return (long)Math.Round(amountMinor * rateBrl * 100m / scale, MidpointRounding.AwayFromZero);
    }

    /// <summary>Conversão com cotação conhecida; sem cotação (moeda estrangeira sem valor salvo) devolve null.</summary>
    public static long? ToBaseCents(long amountMinor, string currency, IReadOnlyDictionary<string, decimal> rates) =>
        currency == Currencies.Base ? amountMinor
        : rates.TryGetValue(currency, out var rate) ? ToBaseCents(amountMinor, currency, rate)
        : null;

    /// <summary>Lê uma cotação decimal (ponto ou vírgula como separador); null quando inválida ou não positiva.</summary>
    public static decimal? ParseRate(string? text)
    {
        if (string.IsNullOrWhiteSpace(text)) return null;
        var normalized = text.Trim();
        if (normalized.Contains(',') && !normalized.Contains('.')) normalized = normalized.Replace(',', '.');
        return decimal.TryParse(normalized, NumberStyles.AllowDecimalPoint | NumberStyles.AllowLeadingSign, Invariant, out var value) && value > 0 ? value : null;
    }

    /// <summary>Texto canônico da cotação (ponto decimal, até 10 casas, sem zeros à direita).</summary>
    public static string FormatRate(decimal rate) => rate.ToString("0.##########", Invariant);

    private static decimal Pow10(int decimals)
    {
        var value = 1m;
        for (var index = 0; index < decimals; index++) value *= 10m;
        return value;
    }
}
